core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProjectOrchestrator.scala

```diff
@@ -1,34 +1,72 @@
+ * 2. Create the rawls internal billing project record
+ * 3. Post-flight steps; this may include syncing of groups, reaching out to external services to sync state, etc.
+ * This step is delegated to the billing project creator as well.
  */
-class BillingProjectOrchestrator(userInfo: UserInfo, samDAO: SamDAO, gcsDAO: GoogleServicesDAO, billingRepository: BillingRepository)(implicit val executionContext: ExecutionContext) extends StringValidationUtils {
+class BillingProjectOrchestrator(userInfo: UserInfo,
+                                 samDAO: SamDAO,
+                                 billingRepository: BillingRepository,
+                                 googleBillingProjectCreator: BillingProjectCreator,
+                                 bpmBillingProjectCreator: BillingProjectCreator)
+                                (implicit val executionContext: ExecutionContext) extends StringValidationUtils with LazyLogging {
   implicit val errorReportSource = ErrorReportSource("rawls")
 
+  /**
+   * Creates a "v2" billing project, using either Azure managed app coordinates or a Google Billing Account
+   */
   def createBillingProjectV2(createProjectRequest: CreateRawlsV2BillingProjectFullRequest): Future[Unit] = {
+    val billingProjectCreator = createProjectRequest.billingInfo match {
+      case Left(_) => googleBillingProjectCreator
+      case Right(_) => bpmBillingProjectCreator
+    }
+    val billingProjectName = createProjectRequest.projectName
+
     for {
       _ <- validateBillingProjectName(createProjectRequest.projectName.value)
-      _ <- ServicePerimeterService.checkServicePerimeterAccess(samDAO, createProjectRequest.servicePerimeter, userInfo)
-      hasAccess <- gcsDAO.testBillingAccountAccess(createProjectRequest.billingAccount, userInfo)
-      _ = if (!hasAccess) {
-        throw new GoogleBillingAccountAccessException(ErrorReport(StatusCodes.BadRequest, "Billing account does not exist, user does not have access, or Terra does not have access"))
+
+      _ = logger.info(s"Validating billing project creation request [name=${billingProjectName.value}]")
+      _ <- billingProjectCreator.validateBillingProjectCreationRequest(createProjectRequest, userInfo)
+
+      _ = logger.info(s"Creating billing project record [name=${billingProjectName}]")
+      _ <- createV2BillingProjectInternal(createProjectRequest, userInfo)
+
+      _ = logger.info(s"Created billing project record, running post-creation steps [name=${billingProjectName.value}]")
+      result <- billingProjectCreator.postCreationSteps(createProjectRequest, userInfo).recoverWith {
+        case t: Throwable =>
+          // rollback any resources we created
+          logger.error(s"Error in post-creation steps for billing project [name=${billingProjectName.value}]")
+          for {
+            _ <-billingRepository.deleteBillingProject(createProjectRequest.projectName).recover {
```

Comment:
These are rolling back things created in `createV2BillingProjectInternal`. I wonder if it's worth putting these in a method called `rollbackCreateV2BillingProjectInternal` for clarity. I really like how the rest of `createBillingProjectV2` reads, but when I got here I was like… wait, why does this code know what  `createV2BillingProjectInternal` did?

Reply:
`createV2BillingProjectInternal` could possibly even return a rollback function that you would call here.

Reply:
Yeah, that works, I'll move this over there.